src/lib/optimizer/strategy/predicate_reordering_rule.cpp

```diff
@@ -26,114 +20,165 @@ bool is_predicate_style_node(const std::shared_ptr<AbstractLQPNode>& node) {
-  caching_cardinality_estimator->guarantee_join_graph(JoinGraph{{input}, {}});
-
-  // Estimate the output cardinalities of each individual predicate on top of the input LQP, i.e., predicates are
-  // estimated independently
-  auto nodes_and_cardinalities = std::vector<std::pair<std::shared_ptr<AbstractLQPNode>, Cardinality>>{};
-  nodes_and_cardinalities.reserve(predicates.size());
+  // Estimate the cardinality of the input node once to cache its estimated statistics. Since we execute the reordering
+  // recursively, we can be sure the plan below does not change anymore and the input estimations can be cached safely.
+  const auto& cardinality_estimator = cost_estimator->cardinality_estimator;
+  cardinality_estimator->estimate_cardinality(input);
+
+  // To order the predicates, we want to favor predicates with lower cost over predicates with higher cost. We estimate
+  // the cost of each individual predicate on top of the input LQP, i.e., predicates are estimated independently. In the
+  // past, we just used the output cardinality. This turned out to be an oversimplification for finding a good order of
+  // scans and joins with alike selectivity, where joins are more expensive in general. Cost is logically estimated
+  // based on input cardinalities and selectivities of predicates (in fact, it simplifies to summing up scanned input
+  // tuples and expected output tuples, see CostEstimatorLogical).
+  //
+  // We experimented with the following optimization goals:
+  //   0) min #out                         (Minimal output cardinality, baseline)
+  //   1) max (#in - #out) / (cost - #out) ("Most filtered-out rows per cost")
+  //   2) min cost                         (Minimal cost)
+  //   3) min #out * cost                  (Minimal output cardinality with cost penalty, also tried +, log, sqrt, ...)
+  //   4) min (cost - out) * p + #out      (Cost with a penalty for joins, chosen approach. p = 1.5 for joins, 1 else.)
+  //
+  // We ended up using 2) the estimated cost to account for the fact that joins are more expensive than predicates.
+  // Furthermore, we add a penalty to the input cardinalities since joins always have more overhead than predicates. The
+  // factor was derived experimentally, which is far from being a perfect solution, but still better than not
+  // incorporating join overhead at all.
+  auto nodes_and_costs = std::vector<std::pair<std::shared_ptr<AbstractLQPNode>, Cost>>{};
+  nodes_and_costs.reserve(predicates.size());
   for (const auto& predicate : predicates) {
     predicate->set_left_input(input);
-    nodes_and_cardinalities.emplace_back(predicate, caching_cardinality_estimator->estimate_cardinality(predicate));
+    // Estimate the cardinality and cost of the predicate without caching. As the predicate order is not yet determined,
+    // caching leads to wrong estimates in the cache.
+    constexpr auto do_cache = false;
```

Comment:
Shouldn't tidy complain and ask for `DO_CACHE`?

Reply:
Interestingly, searching for
```regex
constexpr .+[a-z] =
```

returns
```
138 matches across 48 files
```

Can fix that when we talked about the comments.

Reply:
No need to fix that now (if we want it upper case, it needs to be handled by tidy and tidy will change quite a lot soon).